# StreamFlix Content Analytics Project

## Phase 1 – Data Cleaning & Quality Check

**Analyst:** Hemanth

### Objective
The objective of Phase 1 is to load, understand, validate, and clean the StreamFlix datasets before performing exploratory data analysis and business KPI calculations.

In [2]:
import pandas as pd
import numpy as np

In [3]:
import os

print(os.getcwd())

c:\Users\heman\OneDrive\Desktop\StreamFlix_Content_Analytics_Project\notebooks


In [4]:
subscribers = pd.read_csv("../data/subscribers.csv")

In [5]:
subscribers.head()

,subscriber_id,signup_date,country,region,age,gender,plan_type,monthly_price_usd,household_size,primary_device,payment_method,tenure_months,is_active,churn_date
0,SUB100000,2023-01-09,France,Europe,22,Female,Basic with Ads,6.99,1,Mobile,Credit Card,41,True,NaN
1,SUB100001,2020-02-05,United Kingdom,Europe,48,Male,Basic with Ads,6.99,1,Smart TV,Credit Card,76,False,2022-12-08
2,SUB100002,2022-04-15,Italy,Europe,35,Male,Standard,15.49,2,Streaming Stick,Gift Card,50,True,NaN
3,SUB100003,2024-10-10,Spain,Europe,49,Female,Basic with Ads,6.99,1,Laptop,Credit Card,19,True,NaN
4,SUB100004,2020-04-12,United States,North America,36,Female,Basic with Ads,6.99,1,Mobile,Credit Card,74,False,2021-12-15


In [6]:
subscribers.shape

(15000, 14)

In [7]:
subscribers.columns

Index(['subscriber_id', 'signup_date', 'country', 'region', 'age', 'gender',
       'plan_type', 'monthly_price_usd', 'household_size', 'primary_device',
       'payment_method', 'tenure_months', 'is_active', 'churn_date'],
      dtype='str')

In [8]:
subscribers.info()

<class 'pandas.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   subscriber_id      15000 non-null  str    
 1   signup_date        15000 non-null  str    
 2   country            15000 non-null  str    
 3   region             15000 non-null  str    
 4   age                15000 non-null  int64  
 5   gender             15000 non-null  str    
 6   plan_type          15000 non-null  str    
 7   monthly_price_usd  15000 non-null  float64
 8   household_size     15000 non-null  int64  
 9   primary_device     15000 non-null  str    
 10  payment_method     15000 non-null  str    
 11  tenure_months      15000 non-null  int64  
 12  is_active          15000 non-null  bool   
 13  churn_date         3801 non-null   str    
dtypes: bool(1), float64(1), int64(3), str(9)
memory usage: 1.5 MB


In [9]:
subscribers.isnull().sum()

subscriber_id            0
signup_date              0
country                  0
region                   0
age                      0
gender                   0
plan_type                0
monthly_price_usd        0
household_size           0
primary_device           0
payment_method           0
tenure_months            0
is_active                0
churn_date           11199
dtype: int64

In [10]:
(subscribers.isnull().sum() / len(subscribers)) * 100

subscriber_id         0.00
signup_date           0.00
country               0.00
region                0.00
age                   0.00
gender                0.00
plan_type             0.00
monthly_price_usd     0.00
household_size        0.00
primary_device        0.00
payment_method        0.00
tenure_months         0.00
is_active             0.00
churn_date           74.66
dtype: float64

In [11]:
subscribers['subscriber_id'].duplicated().sum()

np.int64(0)

In [12]:
subscribers.duplicated().sum()

np.int64(0)

In [13]:
subscribers['signup_date'] = pd.to_datetime(subscribers['signup_date'])
subscribers['churn_date'] = pd.to_datetime(subscribers['churn_date'])

In [14]:
subscribers[['signup_date', 'churn_date']].dtypes

signup_date    datetime64[us]
churn_date     datetime64[us]
dtype: object

In [15]:
active_with_churn_date = subscribers[
    (subscribers['is_active'] == True) &
    (subscribers['churn_date'].notna())
]

active_with_churn_date.shape

(0, 14)

In [16]:
inactive_without_churn_date = subscribers[
    (subscribers['is_active'] == False) &
    (subscribers['churn_date'].isna())
]

inactive_without_churn_date.shape

(0, 14)

In [17]:
invalid_churn_dates = subscribers[
    subscribers['churn_date'].notna() &
    (subscribers['churn_date'] <= subscribers['signup_date'])
]

invalid_churn_dates.shape

(0, 14)

In [18]:
subscribers['plan_type'].value_counts()

plan_type
Standard          6100
Basic with Ads    4458
Premium           4442
Name: count, dtype: int64

In [19]:
subscribers['age'].agg(['min', 'max', 'mean'])

min     13.000000
max     85.000000
mean    35.620267
Name: age, dtype: float64

In [20]:
titles = pd.read_csv("../data/titles.csv")

In [21]:
titles.head()

,title_id,title_name,type,primary_genre,country,language,release_year,date_added,maturity_rating,seasons,...,is_original,license_type,director,cast,quality_score,popularity_score,license_cost_usd,license_expiry,total_watch_hours,total_plays
0,TTL200000,Toronto Requiem,TV Show,Comedy,South Korea,Korean,2026,2026-04-05,R,6,...,True,Original,Ravi Nguyen,"Rosa Ferrari, Aria Cohen, Nina Kim, Felix Santos",68.6,0.19,7965521,NaN,1300.8,44
1,TTL200001,Seoul Inheritance,TV Show,Fantasy,South Korea,Korean,2026,2026-05-01,TV-MA,4,...,False,Non-Exclusive License,Greta Sharma,"Priya Ferrari, Lucia Mehta, Ravi Yilmaz, Joon ...",89.4,0.29,3074516,2028-08-03,1767.2,52
2,TTL200002,Tender Paradox,TV Show,Horror,Italy,Italian,2026,2026-04-17,PG,3,...,False,Non-Exclusive License,Bruno Moreau,"Elena Park, Aria Romano, Omar Olsen, Chloe Nak...",37.3,0.20,629307,2027-02-16,166.5,60
3,TTL200003,Velvet Expedition,Movie,Animation,India,Hindi,1985,2018-12-08,TV-14,0,...,False,Non-Exclusive License,Owen Mwangi,"Ava Russo, Clara Fischer, Zoe Khan, Rosa Reyes",37.8,0.02,1345769,2029-10-11,16.2,43
4,TTL200004,The Shadows of Lisbon,Movie,Animation,India,Hindi,1991,2016-05-11,TV-14,0,...,False,Exclusive License,Joon Kim,"Hana Patel, Fatima Patel, Hassan Reyes, Liam L...",78.0,0.00,1696493,2028-11-21,35.7,43


In [22]:
titles.shape

(9000, 21)

In [23]:
titles.columns

Index(['title_id', 'title_name', 'type', 'primary_genre', 'country',
       'language', 'release_year', 'date_added', 'maturity_rating', 'seasons',
       'content_duration_min', 'is_original', 'license_type', 'director',
       'cast', 'quality_score', 'popularity_score', 'license_cost_usd',
       'license_expiry', 'total_watch_hours', 'total_plays'],
      dtype='str')

In [24]:
titles.info()

<class 'pandas.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   title_id              9000 non-null   str    
 1   title_name            9000 non-null   str    
 2   type                  9000 non-null   str    
 3   primary_genre         9000 non-null   str    
 4   country               9000 non-null   str    
 5   language              9000 non-null   str    
 6   release_year          9000 non-null   int64  
 7   date_added            9000 non-null   str    
 8   maturity_rating       9000 non-null   str    
 9   seasons               9000 non-null   int64  
 10  content_duration_min  9000 non-null   int64  
 11  is_original           9000 non-null   bool   
 12  license_type          9000 non-null   str    
 13  director              9000 non-null   str    
 14  cast                  9000 non-null   str    
 15  quality_score         9000 non-n

In [25]:
titles.isnull().sum()

title_id                   0
title_name                 0
type                       0
primary_genre              0
country                    0
language                   0
release_year               0
date_added                 0
maturity_rating            0
seasons                    0
content_duration_min       0
is_original                0
license_type               0
director                   0
cast                       0
quality_score              0
popularity_score           0
license_cost_usd           0
license_expiry          1966
total_watch_hours          0
total_plays                0
dtype: int64

In [26]:
(titles.isnull().sum() / len(titles)) * 100

title_id                 0.000000
title_name               0.000000
type                     0.000000
primary_genre            0.000000
country                  0.000000
language                 0.000000
release_year             0.000000
date_added               0.000000
maturity_rating          0.000000
seasons                  0.000000
content_duration_min     0.000000
is_original              0.000000
license_type             0.000000
director                 0.000000
cast                     0.000000
quality_score            0.000000
popularity_score         0.000000
license_cost_usd         0.000000
license_expiry          21.844444
total_watch_hours        0.000000
total_plays              0.000000
dtype: float64

In [27]:
titles[titles['license_expiry'].isna()]['is_original'].value_counts()

is_original
True    1966
Name: count, dtype: int64

In [28]:
titles['title_id'].duplicated().sum()

np.int64(0)

In [29]:
titles.duplicated().sum()

np.int64(0)

In [30]:
titles['date_added'] = pd.to_datetime(titles['date_added'])
titles['license_expiry'] = pd.to_datetime(titles['license_expiry'])

In [31]:
titles[['date_added', 'license_expiry']].dtypes

date_added        datetime64[us]
license_expiry    datetime64[us]
dtype: object

In [32]:
titles['type'].value_counts()

type
Movie      5336
TV Show    3664
Name: count, dtype: int64

In [33]:
movies_with_seasons = titles[
    (titles['type'] == 'Movie') &
    (titles['seasons'] != 0)
]

movies_with_seasons.shape

(0, 21)

In [34]:
tvshows_without_seasons = titles[
    (titles['type'] == 'TV Show') &
    (titles['seasons'] <= 0)
]

tvshows_without_seasons.shape

(0, 21)

In [35]:
watch_history = pd.read_csv("../data/watch_history.csv")

In [36]:
watch_history = pd.read_csv("../data/watch_history.csv")

In [37]:
watch_history.head()

,watch_id,subscriber_id,title_id,watch_date,device,region,content_duration_min,watch_duration_min,completion_pct,completed
0,1,SUB107630,TTL205207,2024-01-01,Smart TV,East Asia,888,826.7,93.1,True
1,2,SUB114773,TTL204889,2025-10-18,Mobile,South Asia,2340,1866.4,79.8,False
2,3,SUB106013,TTL206882,2026-04-26,Mobile,Latin America,656,75.0,11.4,False
3,4,SUB100939,TTL208849,2026-03-09,Mobile,North America,500,493.2,98.6,True
4,5,SUB107223,TTL202529,2023-12-27,Tablet,North America,141,94.6,67.1,False


In [38]:
watch_history.shape

(650000, 10)

In [39]:
watch_history.info()

<class 'pandas.DataFrame'>
RangeIndex: 650000 entries, 0 to 649999
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   watch_id              650000 non-null  int64  
 1   subscriber_id         650000 non-null  str    
 2   title_id              650000 non-null  str    
 3   watch_date            650000 non-null  str    
 4   device                650000 non-null  str    
 5   region                650000 non-null  str    
 6   content_duration_min  650000 non-null  int64  
 7   watch_duration_min    650000 non-null  float64
 8   completion_pct        650000 non-null  float64
 9   completed             650000 non-null  bool   
dtypes: bool(1), float64(2), int64(2), str(5)
memory usage: 45.3 MB


In [40]:
watch_history['watch_date'] = pd.to_datetime(watch_history['watch_date'])

In [41]:
watch_history['watch_date'].dtype

dtype('<M8[us]')

In [42]:
invalid_watch_duration = watch_history[
    watch_history['watch_duration_min'] > watch_history['content_duration_min']
]

invalid_watch_duration.shape

(0, 10)

In [43]:
watch_history['watch_id'].duplicated().sum()

np.int64(0)

In [44]:
watch_history.duplicated().sum()

np.int64(0)

In [45]:
completion_difference = abs(
    watch_history['completion_pct'] -
    (watch_history['watch_duration_min'] / watch_history['content_duration_min'] * 100)
)

(completion_difference > 0.1).sum()

np.int64(1845)

In [46]:
completion_difference.describe()

count    650000.000000
mean          0.029353
std           0.021061
min           0.000000
25%           0.012500
50%           0.026531
75%           0.042373
max           0.155556
dtype: float64

In [47]:
calculated_completion = (
    watch_history['watch_duration_min'] /
    watch_history['content_duration_min'] * 100
).round(1)

(calculated_completion != watch_history['completion_pct']).sum()

np.int64(100507)

In [48]:
(completion_difference > 0.2).sum()

np.int64(0)

In [49]:
invalid_subscriber_ids = watch_history[
    ~watch_history['subscriber_id'].isin(subscribers['subscriber_id'])
]

invalid_subscriber_ids.shape

(0, 10)

In [50]:
invalid_title_ids = watch_history[
    ~watch_history['title_id'].isin(titles['title_id'])
]

invalid_title_ids.shape

(0, 10)

In [51]:
ratings = pd.read_csv("../data/ratings.csv")

In [52]:
ratings.head()

,rating_id,subscriber_id,title_id,rating,rating_date
0,1,SUB109184,TTL203909,4,2026-02-06
1,2,SUB105194,TTL200888,4,2025-03-21
2,3,SUB102974,TTL200036,3,2024-05-26
3,4,SUB105612,TTL206358,5,2026-05-22
4,5,SUB103070,TTL205227,4,2026-05-16


In [53]:
ratings.shape


(130000, 5)

In [54]:
ratings.info()

<class 'pandas.DataFrame'>
RangeIndex: 130000 entries, 0 to 129999
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype
---  ------         --------------   -----
 0   rating_id      130000 non-null  int64
 1   subscriber_id  130000 non-null  str  
 2   title_id       130000 non-null  str  
 3   rating         130000 non-null  int64
 4   rating_date    130000 non-null  str  
dtypes: int64(2), str(3)
memory usage: 5.0 MB


In [55]:
ratings['rating_date'] = pd.to_datetime(ratings['rating_date'])

In [56]:
ratings['rating_date'].dtype

dtype('<M8[us]')

In [57]:
ratings['rating'].value_counts().sort_index()

rating
1     1786
2    14570
3    44491
4    48564
5    20589
Name: count, dtype: int64

In [58]:
ratings['rating_id'].duplicated().sum()

np.int64(0)

In [59]:
ratings.duplicated().sum()

np.int64(0)

In [60]:
invalid_rating_subscribers = ratings[
    ~ratings['subscriber_id'].isin(subscribers['subscriber_id'])
]

invalid_rating_subscribers.shape

(0, 5)

In [61]:
invalid_rating_titles = ratings[
    ~ratings['title_id'].isin(titles['title_id'])
]

invalid_rating_titles.shape

(0, 5)

In [62]:
reviews = pd.read_csv("../data/reviews.csv")

In [63]:
reviews.head()

,review_id,subscriber_id,title_id,review_text,sentiment,helpful_votes,review_date
0,1,SUB102409,TTL206425,Stellar cast and a satisfying ending.,Positive,9,2023-11-04
1,2,SUB111270,TTL203987,Predictable plot and flat characters.,Negative,4,2026-04-25
2,3,SUB107848,TTL203235,"Started strong, lost steam in the middle.",Neutral,2,2024-07-31
3,4,SUB114501,TTL200930,Beautifully shot and emotionally resonant.,Positive,14,2022-12-13
4,5,SUB113370,TTL207141,A rare gem. Highly recommend.,Positive,6,2025-10-09


In [64]:
reviews.shape

(110000, 7)

In [65]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 110000 entries, 0 to 109999
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype
---  ------         --------------   -----
 0   review_id      110000 non-null  int64
 1   subscriber_id  110000 non-null  str  
 2   title_id       110000 non-null  str  
 3   review_text    110000 non-null  str  
 4   sentiment      110000 non-null  str  
 5   helpful_votes  110000 non-null  int64
 6   review_date    110000 non-null  str  
dtypes: int64(2), str(5)
memory usage: 5.9 MB


In [66]:
reviews['review_date'] = pd.to_datetime(reviews['review_date'])


In [67]:
reviews['review_date'].dtype

dtype('<M8[us]')

In [68]:
reviews['sentiment'].value_counts()

sentiment
Positive    75007
Neutral     22759
Negative    12234
Name: count, dtype: int64

In [69]:
reviews['review_id'].duplicated().sum()

np.int64(0)

In [70]:
reviews.duplicated().sum()

np.int64(0)

In [71]:
(reviews['helpful_votes'] < 0).sum()

np.int64(0)

In [72]:
invalid_review_subscribers = reviews[
    ~reviews['subscriber_id'].isin(subscribers['subscriber_id'])
]

invalid_review_subscribers.shape

(0, 7)

In [73]:
invalid_review_titles = reviews[
    ~reviews['title_id'].isin(titles['title_id'])
]

invalid_review_titles.shape

(0, 7)

In [74]:
watchlist = pd.read_csv("../data/watchlist.csv")

In [75]:
watchlist.head()

,watchlist_id,subscriber_id,title_id,added_date,watched
0,1,SUB104004,TTL201827,2023-06-05,True
1,2,SUB102310,TTL201354,2023-01-31,False
2,3,SUB111860,TTL204871,2022-06-21,True
3,4,SUB113333,TTL205130,2023-12-23,True
4,5,SUB106850,TTL204396,2022-02-02,False


In [76]:
watchlist.shape

(65000, 5)

In [77]:
watchlist.info()

<class 'pandas.DataFrame'>
RangeIndex: 65000 entries, 0 to 64999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   watchlist_id   65000 non-null  int64
 1   subscriber_id  65000 non-null  str  
 2   title_id       65000 non-null  str  
 3   added_date     65000 non-null  str  
 4   watched        65000 non-null  bool 
dtypes: bool(1), int64(1), str(3)
memory usage: 2.0 MB


In [78]:
watchlist['added_date'] = pd.to_datetime(watchlist['added_date'])

In [79]:
watchlist['added_date'].dtype

dtype('<M8[us]')

In [80]:
watchlist['watchlist_id'].duplicated().sum()

np.int64(0)

In [81]:
watchlist.duplicated().sum()

np.int64(0)

In [82]:
invalid_watchlist_subscribers = watchlist[
    ~watchlist['subscriber_id'].isin(subscribers['subscriber_id'])
]

invalid_watchlist_subscribers.shape

(0, 5)

In [83]:
invalid_watchlist_titles = watchlist[
    ~watchlist['title_id'].isin(titles['title_id'])
]

invalid_watchlist_titles.shape

(0, 5)

## Phase 1 – Data Quality Summary

### 1. subscribers.csv
- Total records: 15,000
- Total columns: 14
- Duplicate subscriber IDs: 0
- Fully duplicated rows: 0
- Missing values were found only in `churn_date`.
- `churn_date` has 11,199 missing values, representing 74.66% of the table.
- The missing churn dates are expected for active subscribers.
- No active subscribers had a churn date.
- No inactive subscribers were missing a churn date.
- All churn dates occurred after the corresponding signup dates.
- `signup_date` and `churn_date` were converted to datetime format.
- Subscriber ages ranged from 13 to 85 years.
- The average subscriber age was approximately 35.62 years.
- All plan types were valid: Basic with Ads, Standard, and Premium.

### 2. titles.csv
- Total records: 9,000
- Total columns: 21
- Duplicate title IDs: 0
- Fully duplicated rows: 0
- Missing values were found only in `license_expiry`.
- `license_expiry` has 1,966 missing values, representing approximately 21.84% of the table.
- All missing licence-expiry values belonged to StreamFlix Original titles, so these null values are valid.
- `date_added` and `license_expiry` were converted to datetime format.
- The dataset contains 5,336 Movies and 3,664 TV Shows.
- All Movies correctly have `seasons = 0`.
- All TV Shows have at least one season.

### 3. watch_history.csv
- Total records: 650,000
- Total columns: 10
- Missing values: 0
- Duplicate watch IDs: 0
- Fully duplicated rows: 0
- `watch_date` was converted to datetime format.
- No sessions had `watch_duration_min` greater than `content_duration_min`.
- `completion_pct` was consistent with the calculated completion percentage.
- Small differences were caused by rounding.
- The maximum completion-percentage difference was approximately 0.16 percentage points.
- No material completion-percentage differences greater than 0.2 percentage points were found.
- Every subscriber ID exists in the subscribers table.
- Every title ID exists in the titles table.

### 4. ratings.csv
- Total records: 130,000
- Total columns: 5
- Missing values: 0
- Duplicate rating IDs: 0
- Fully duplicated rows: 0
- `rating_date` was converted to datetime format.
- All ratings were within the valid 1–5 range.
- Every subscriber ID exists in the subscribers table.
- Every title ID exists in the titles table.

### 5. reviews.csv
- Total records: 110,000
- Total columns: 7
- Missing values: 0
- Duplicate review IDs: 0
- Fully duplicated rows: 0
- `review_date` was converted to datetime format.
- Sentiment values contained only Positive, Neutral, and Negative.
- No negative helpful-vote counts were found.
- Every subscriber ID exists in the subscribers table.
- Every title ID exists in the titles table.

### 6. watchlist.csv
- Total records: 65,000
- Total columns: 5
- Missing values: 0
- Duplicate watchlist IDs: 0
- Fully duplicated rows: 0
- `added_date` was converted to datetime format.
- Every subscriber ID exists in the subscribers table.
- Every title ID exists in the titles table.

### Overall Data Quality Conclusion
The StreamFlix datasets are generally high quality. No duplicate primary keys, full duplicate rows, invalid foreign-key relationships, or major logical inconsistencies were identified. Missing values in `churn_date` and `license_expiry` are expected based on the business rules of the dataset. Date columns were converted to proper datetime formats, and the cleaned datasets are ready for exploratory data analysis and KPI calculations.

In [86]:
subscribers.to_csv("../cleaned_data/subscribers_cleaned.csv", index=False)
titles.to_csv("../cleaned_data/titles_cleaned.csv", index=False)
watch_history.to_csv("../cleaned_data/watch_history_cleaned.csv", index=False)
ratings.to_csv("../cleaned_data/ratings_cleaned.csv", index=False)
reviews.to_csv("../cleaned_data/reviews_cleaned.csv", index=False)
watchlist.to_csv("../cleaned_data/watchlist_cleaned.csv", index=False)

print("All cleaned datasets saved successfully.")

All cleaned datasets saved successfully.
